dotnet/src/Connectors/Connectors.Memory.Milvus/MilvusMemoryStore.cs

```diff
@@ -0,0 +1,429 @@
+
+        if (await enumerator.MoveNextAsync().ConfigureAwait(false) == false)
+        {
+            return null;
+        }
+
+        (MemoryRecord, double) result = enumerator.Current;
+
+        Debug.Assert(await enumerator.MoveNextAsync().ConfigureAwait(false) == false);
+
+        return result;
+    }
+
+    public async IAsyncEnumerable<(MemoryRecord, double)> GetNearestMatchesAsync(
+        string collectionName,
+        ReadOnlyMemory<float> embedding,
+        int limit,
+        double minRelevanceScore = 0,
+        bool withEmbeddings = false,
+        CancellationToken cancellationToken = default)
+    {
+        MilvusCollection collection = this._milvusClient.GetCollection(collectionName);
+
+        SearchResults results = await collection
+            .SearchAsync(EmbeddingFieldName, new[] { embedding }, SimilarityMetricType.Ip, limit, this._searchParameters, cancellationToken)
+            .ConfigureAwait(false);
+
+        IReadOnlyList<string> ids = results.Ids.StringIds!;
+        int rowCount = ids.Count;
+        IReadOnlyList<FieldData> data = results.FieldsData;
+
+        // Since Milvus does not support fetching vectors via the Search API, we do an extra call to fetch the ids and embeddings using the Query API,
+        // using the IDs returned from the Search above, populating a map from the IDs to the embedding.
+        Dictionary<string, ReadOnlyMemory<float>>? embeddingMap = null;
+        if (withEmbeddings)
+        {
+            StringBuilder filter = new();
+            filter.Append(IdFieldName).Append(" in [");
+
+            for (int rowNum = 0; rowNum < ids.Count; rowNum++)
+            {
+                if (rowNum > 0)
+                {
+                    filter.Append(',');
+                }
+
+                filter.Append('"').Append(ids[rowNum]).Append('"');
+            }
+
+            filter.Append(']');
+
+            IReadOnlyList<FieldData> fieldData = await collection.QueryAsync(
+                    filter.ToString(),
+                    new() { OutputFields = { EmbeddingFieldName } },
+                    cancellationToken: cancellationToken)
+                .ConfigureAwait(false);
+
+            IReadOnlyList<string> idData = (fieldData[0] as FieldData<string> ?? fieldData[1] as FieldData<string>)!.Data;
+            IReadOnlyList<ReadOnlyMemory<float>> embeddingData = (fieldData[0] as FloatVectorFieldData ?? fieldData[1] as FloatVectorFieldData)!.Data;
+
+            embeddingMap = new Dictionary<string, ReadOnlyMemory<float>>(ids.Count);
+            for (int rowNum = 0; rowNum < ids.Count; rowNum++)
+            {
+                embeddingMap[idData[rowNum]] = embeddingData[rowNum];
+            }
+        }
+
+        for (int rowNum = 0; rowNum < rowCount; rowNum++)
+        {
+            if (results.Scores[rowNum] >= minRelevanceScore)
```

Comment:
Nothing in the search API allows for such a threshold to be respected on the db side?

Reply:
Not as far as I can tell - and this actually is pretty problematic in its interactive with limit, which **is** processed db-side (see https://github.com/microsoft/semantic-kernel/pull/1941/files#r1313978902).

I'll reach out to the Milvus people to make sure there's nothing we can use on their end.

Reply:
Update: the recently-released Milvus 2.3 ([release notes](https://github.com/milvus-io/milvus/releases/tag/v2.3.0)) has a new feature - [range search](https://milvus.io/docs/within_range.md#Conduct-a-Range-Search) - which allows doing this. This would be considerably better than the current client-side implementation.

/cc @awharrison-28